lib/spack/docs/contribution_guide.rst

```diff
@@ -0,0 +1,242 @@
+
+Important is completed feature, i.e. the changes one propose in a PR should
+correspond to one feature/bugfix/extension/etc. One can create PRs with
+changes relevant to different ideas, however reviewing such PRs becomes tedious
+and error prone. If possible, try to follow the rule **one-PR-one-package/feature.**
+
+
+Let's assume that the current (patched) state of your fork of Spack is only
+relevant to yourself. Now you come across a bug in a package or would like to
+extend a package and contribute this fix to Spack. It is important that
+whenever you change something that might be of importance upstream,
+create a pull-request (PR) as soon as possible. Do not wait for weeks/months to
+do this: a) you might forget why did you modified certain files; b) it could get
+difficult to isolate this change into a stand-alone clean PR.
+
+Now let us discuss several approaches one may use to submit a PR while
+also keeping your local version of Spack patched.
+
+
+First approach (cherry-picking):
+--------------------------------
+
+First approach is as follows.
+You checkout your local develop branch which for the purpose of this guide
+will be called ``develop_modified``:
+
+.. code-block:: console
+
+  $ git checkout develop_modified
+
+Let us assume that lines in files you will be modifying
+are the same in `develop_modified` branch and upstream ``develop``.
+Next edit files, make sure they work for you and create a commit
+
+.. code-block:: console
+
+  $ git add <files_to_be_commited>
+  $ git commit -m <descriptive note about changes>
+
+Normally we prefer a commits pertaining to a package ``<package-name>``` have
+a message ``<package-name>: descriptive message``. It is important to add
+descriptive message so that others, who might be looking at your changes later
+(in a year or maybe two), would understand the rationale behind.
+
+
+Next we will create a branch off upstream's ``develop`` and copy this commit.
+Before doing this, while still on your modified branch, get the hash of the last
+ commit
+
+.. code-block:: console
+
+  $ git log -1
+
+and copy-paste this ``<hash>`` to the buffer. Now switch to upstream's ``develop``,
+make sure it's updated, checkout the new branch, apply the patch and push to
+GitHub:
+
+.. code-block:: console
+
+  $ git checkout develop
+  $ git pull
+  $ git checkout -b <descriptive_branch_name>
+  $ git cherry-pick <hash>
+  $ git push <your_origin> -u
+
+Here we assume that local ``develop`` branch tracks upstream develop branch of
+Spack. This is not a requirement and you could also do the same with remote
+branches. Yet to some it more convenient to have a local branch which
+tracks upstream.
+
+Now you can create a PR from web-interface of GitHub. The net result is as
+follows:
+
+1. You patched your local version of Spack and can use it further
+2. You "cherry-picked" these changes in a stand-alone branch and submitted it
+   as a PR upstream.
```

Comment:
You can use `#.` instead of `1.`, `2.`, etc and Sphinx will automatically replace them with numbers. This makes it easy to later rearrange numbers or add a new number in between.